soroban-env-host/src/e2e_testutils.rs

```diff
@@ -46,31 +50,60 @@ pub fn bytes_sc_val(bytes: &[u8]) -> ScVal {
     ScVal::Bytes(ScBytes(bytes.try_into().unwrap()))
 }
 
-pub fn wasm_entry(wasm: &[u8]) -> LedgerEntry {
+/// Helper for creating a non-validated Wasm entry without refined
+/// cost inputs.
+/// Useful for tests that don't care about the Wasm entry contents.
+pub fn wasm_entry_non_validated(wasm: &[u8]) -> LedgerEntry {
+    let ext = ContractCodeEntryExt::V0;
+
     ledger_entry(LedgerEntryData::ContractCode(ContractCodeEntry {
-        ext: ContractCodeEntryExt::V0,
+        ext,
         hash: get_wasm_hash(wasm).try_into().unwrap(),
         code: wasm.try_into().unwrap(),
     }))
 }
 
-pub fn wasm_entry_with_refined_contract_cost_inputs(wasm: &[u8]) -> LedgerEntry {
-    use crate::xdr::ContractCodeEntryV1;
-    let host = crate::Host::default();
-    ledger_entry(LedgerEntryData::ContractCode(ContractCodeEntry {
-        ext: ContractCodeEntryExt::V1(ContractCodeEntryV1 {
-            ext: ExtensionPoint::V0,
-            cost_inputs: crate::vm::ParsedModule::extract_refined_contract_cost_inputs(&host, wasm)
+/// Creates a ContractCode ledger entry containing the provided Wasm blob.
+/// The entry has the most 'up-to-date' contents possible (such as refined costs
+/// in protocol 21).
+pub fn wasm_entry(wasm: &[u8]) -> LedgerEntry {
+    wasm_entry_with_refined_contract_cost_inputs(
+        wasm,
+        e2e_test_protocol_version() >= ModuleCache::MIN_LEDGER_VERSION,
+    )
+}
+
+pub(crate) fn wasm_entry_with_refined_contract_cost_inputs(
+    wasm: &[u8],
+    _add_refined_cost_inputs: bool,
+) -> LedgerEntry {
+    let ext = if !_add_refined_cost_inputs {
+        ContractCodeEntryExt::V0
+    } else {
+        let dummy_host = crate::Host::test_host();
+        dummy_host.set_ledger_info(default_ledger_info()).unwrap();
+        ContractCodeEntryExt::V1(ContractCodeEntryV1 {
+            cost_inputs: ParsedModule::extract_refined_contract_cost_inputs(&dummy_host, wasm)
                 .unwrap(),
-        }),
+            ext: ExtensionPoint::V0,
+        })
+    };
+    ledger_entry(LedgerEntryData::ContractCode(ContractCodeEntry {
+        ext,
         hash: get_wasm_hash(wasm).try_into().unwrap(),
         code: wasm.try_into().unwrap(),
     }))
 }
 
+pub fn e2e_test_protocol_version() -> u32 {
+    env::var("TEST_PROTOCOL")
+        .map(|v| u32::from_str(&v).unwrap())
+        .unwrap_or(20)
```

Comment:
I don't have a strong opinion here, but I am curious if defaulting to 20 is what you want/need here. The `Host::current_test_protocol` method returns the env's wired-in max supported protocol if `TEST_PROTOCOL` is not set.

(For that matter, can you call Host::current_test_protocol directly here? I have its logic replicated in synth-wasm because it's a separate crate and the dependencies go the other way, but this is part of env-host..)

Reply:
That's actually just a leftover from the previous iterations. I would prefer to still have a separate public function in order to be able to reuse it in simulation and also just in case if we ever want to decouple the protocol versions for different parts of the test suite.